src/memoize.js

```diff
@@ -10,16 +10,35 @@ import type {Context} from './types.js';
 
 type MemoizeFn<A> = (ctx: Context) => A;
 
-function Container() {}
-
 export function memoize<A>(fn: MemoizeFn<A>): MemoizeFn<A> {
-  const memoizeKey = __NODE__ ? Symbol('memoize-key') : new Container();
-  return function memoized(ctx: Context) {
-    if (ctx.memoized.has(memoizeKey)) {
-      return ctx.memoized.get(memoizeKey);
+  if (__BROWSER__) {
+    return browserMemoize(fn);
+  }
+
+  const wm = new WeakMap();
+  return ctx => {
+    if (wm.has(ctx)) {
+      return ((wm.get(ctx): any): A); // Refinement doesn't seem to work
+    } else {
+      const result = fn(ctx);
+      wm.set(ctx, result);
+      return result;
+    }
+  };
+}
+
+/**
+ * There is only ever a single ctx object in the browser.
+ * Therefore we can use a simple memoization function.
```

Comment:
I'm not totally sure if this assumption is valid. This is certainly true in real cases, but I wonder if this may not be true with simulation tests and whatnot.

But if this assumption is not valid, we should add a test case for this.